cpp/pagerank_module/algorithm/pagerank.hpp

```diff
@@ -0,0 +1,181 @@
+/// are [0, 3>, [3, 6> and [6, 10>.
+///
+/// @param graph -- graph
+/// @param number_of_threads -- number of threads
+/// @param borders -- vector in which calculated optimal borders will be written
+void CalcualteOptimalBorders(const PageRankGraph &graph, int number_of_threads, std::vector<std::uint64_t> *borders);
+
+/// Calculating PageRank block related to [lo, hi> interval of edges
+/// Graph edges are ordered by target node ids so target nodes of an interval
+/// of edges also form an interval of nodes.
+/// Required PageRank values of nodes from this interval will be sent
+/// as vector of values using new_rank_promise.
+///
+/// @param graph -- graph
+/// @param old_rank -- rank of nodes after previous iterations
+/// @param damping_factor -- damping factor
+/// @param lo -- left bound of interval of edges
+/// @param hi -- right bound of interval of edges
+/// @param new_rank_promise -- used for sending information about calculated new
+/// PageRank block
+void ThreadPageRankIteration(const PageRankGraph &graph, const std::vector<double> &old_rank, int lo, int hi,
+                             std::promise<std::vector<double>> new_rank_promise);
+
+/// Merging PageRank blocks from ThreadPageRankIteration.
+///
+/// @param graph -- graph
+/// @param damping_factor -- damping factor
+/// @param block -- PageRank block calculated in ThreadPageRankIteration
+/// @param rank_next -- PageRanks which will be updated
+void AddCurrentBlockToRankNext(const PageRankGraph &graph, double damping_factor, const std::vector<double> &block,
+                               std::vector<double> &rank_next);
+
+/// Adds remaining PageRank values
+/// Adds PageRank values of nodes that haven't been added by
+/// AddCurrentBlockToRankNext. That are nodes whose id is greater
+/// than id of target node of the last edge in ordered edge list.
+///
+/// @param graph -- graph
+/// @param damping_factor -- damping factor
+/// @param rank_next -- PageRank which will be updated
+void CompleteRankNext(const PageRankGraph &graph, double damping_factor, std::vector<double> &rank_next);
+
+/// Checks whether PageRank algorithm should continue iterating
+/// Checks if maximal number of iterations was reached or
+/// if difference between every component of previous and current PageRank
+/// was less than stop_epsilon.
+///
+/// @param rank -- PageRank before the last iteration
+/// @param rank_next -- PageRank after the last iteration
+/// @param max_iterations -- maximal number of iterations
+/// @param stop_epsilon -- stop epsilon
+/// @param number_of_iterations -- current number of operations
+bool CheckContinueIterate(const std::vector<double> &rank, const std::vector<double> &rank_next, size_t max_iterations,
+                          double stop_epsilon, size_t number_of_iterations);
+
+/// Normalizing PageRank
+/// Divides all values with sum of the values to get their sum equal 1
+///
+/// @param rank -- PageRank
+void NormalizeRank(std::vector<double> &rank);
```

Comment:
Why are these functions even exposed in the header?
Why not have them in the anonymous namespace of the `pagerank.cpp`.

Reply:
That is much better when looking from the design perspective. I expected everything should be defined in the header file as it is some kind of C++ standard convention.

Reply:
Changes should be visible now